terraform/main.tf

```diff
@@ -0,0 +1,175 @@
+# whether it's intended for a staging or prod environment, so we leave the
+# naming logic up to the Terraform caller, which we expect to be a CI
+# environment with knowledge of the current git branch name
+variable "batch_job_name" {
+  type = string
+}
+
+# Set the name of the container image that the Batch job should pull to use as
+# its execution environment, e.g. "ghcr.io/ccao-data/model-res-avm:master".
+# This is defined as a variable so that CI environments can point Batch
+# job definitions to freshly built images
+variable "batch_container_image_name" {
+  type = string
+}
+
+# Retrieve the default VPC for this region, which is builtin to AWS.
+# Containers in the Batch compute environment will be deployed into this VPC
+data "aws_vpc" "default" {
+  default = true
+}
+
+# Retrieve the default subnets in the default VPC, which are builtin to AWS.
+# Containers in the Batch compute environment are connected to the Internet
+# using these subnets. Note that these subnets are public by default, but the
+# use of an additional security group ensures that we block all ingress to the
+# compute environment
+data "aws_subnets" "default" {
+  filter {
+    name   = "vpc-id"
+    values = [data.aws_vpc.default.id]
+  }
+  filter {
+    name   = "default-for-az"
+    values = [true]
+  }
+  filter {
+    name   = "state"
+    values = ["available"]
+  }
+}
+
+# Retrieve the security group that blocks all ingress and allows egress over
+# HTTPS only
+data "aws_security_group" "outbound_https" {
+  name = "outbound-https"
+  vpc_id = data.aws_vpc.default.id
+}
+
+# Retrieve the IAM role that the Batch compute environment uses to manage
+# EC2/ECS resources
+data "aws_iam_role" "batch_service_role" {
+  name = "AWSServiceRoleForBatch"
+}
+
+# Retrieve the IAM role that the Batch job definition uses to execute ECS
+# operations like pulling Docker images and pushing logs to CloudWatch
+data "aws_iam_role" "ecs_task_execution_role" {
+  name = "ecsTaskExecutionRole"
+}
+
+# Retrieve the IAM role that the Batch job passes on to the containers, allowing
+# those containers to access AWS resources like data stored in S3
+data "aws_iam_role" "ecs_job_role" {
+  name = "ccao-ecs-dvc-full-access"
+}
+
+# Create a Batch compute environment to run containers. Compute environments
+# define the underlying ECS or EC2 resources that will be provisioned to
+# use for running jobs in containers. Docs here:
+# https://docs.aws.amazon.com/batch/latest/userguide/compute_environments.html
+resource "aws_batch_compute_environment" "main" {
+  compute_environment_name = var.batch_job_name
+  service_role             = data.aws_iam_role.batch_service_role.arn
+  state                    = "ENABLED"
+  type                     = "MANAGED"
+
+  compute_resources {
+    type               = "FARGATE"
+    min_vcpus          = 0
+    max_vcpus          = 64  # Max across all jobs, not within one job
```

Comment:
For example, if our jobs require 16 cores, then there can be a maximum of 64/16 = 4 jobs running at once time in this compute environment; jobs requiring CPU beyond that limit will be queued. The current design of the workflow that uses these environments is not particularly well designed to handle queueing, since we configure jobs to fail if they don't startup in 10 minutes, but these are variables that we can easily tweak in the future if it becomes a problem.

Reply:
Makes sense! Not too worried about queuing since we manually approve each run AND each PR has its own compute environment.

This also mostly answers my earlier question.

Reply:
I actually thought your earlier question was a reasonable one! In theory, it should be straightforward for all jobs of a given type (e.g. res model) to share a compute environment and job queue. The reason I went with isolation here is because I'm expecting this configuration will be reused across repos, and while we can expect that subnets, security groups, and IAM roles will be stable and reusable across repos, compute environments and job queues will likely require different configurations for different repos. While we could reduce the number of resources we create and destroy by managing these resources on a repo level, I think it's helpful to build staging versions in case we decide we want to change an environment/queue configuration and we want to test the change before deploying it.

Reply:
Totally makes sense and I like the flexibility to being able to change environment per repo _and_ per branch. Nice thinking 👍